src/braket/experimental/autoqasm/program/program.py

```diff
@@ -322,22 +323,21 @@ def register_args(self, args: list[Any]) -> None:
         """
         for arg in args:
             if isinstance(arg, FreeParameter):
-                self.register_parameter(arg)
+                self.register_parameter(arg.name)
             elif isinstance(arg, FreeParameterExpression):
-                # TODO laurecap: Support for expressions
-                raise NotImplementedError(
-                    "Expressions of FreeParameters will be supported shortly!"
-                )
+                for expression_arg in preorder_traversal(arg._expression.args):
+                    if isinstance(expression_arg, Symbol):
+                        self.register_parameter(expression_arg.name)
 
-    def register_parameter(self, parameter: FreeParameter) -> None:
+    def register_parameter(self, parameter_name: str) -> None:
```

Comment:
Could this accept a union of str, FreeParam?

Then L339, 340 can check for repr(parameter_name) and add that value in.